content/blog/2019-10-23_react-typescript-starter.md

```diff
@@ -0,0 +1,464 @@
+- `target`: we target modern browsers so ES6 is fine here. If you need IE11 support, change it to `es5`
+- `jsx`: we use React and transpile from TypeScript directly so we use `react`
+- `sourceMap`: we want them
+- `strict`: enable all the strict flags automatically
+
+All the other options are adding some additional checks, such as not allowing unused variables and a list 
+of all the options can be found in [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/compiler-options.html).
+Lastly, we set up an include path containing `"./src/**/*"`: we will put all our code in a `src` folder.
+
+In this setup, we will only use TypeScript and not [Babel](https://babeljs.io/). Is it possible to use TypeScript for
+typechecking only and pass the code to Babel for transpilation but, unless you need a specific Babel plugin, there are no 
+reasons to do that nowadays. You will have to wait a few more months before getting stage 3 features - not a big deal.
+
+The commit for that step can be seen in <https://github.com/Keats/webpack-react-typescript/commit/a083ce1511b15c51d4a8d3ef95f2da05a669758b>.
+
+## Webpack setup
+Let's start the slightly more complex part! Once you understand how Webpack works, this is actually pretty simple.
+
+### Webpack itself
+We first need webpack and its CLI:
+
+```bash
+$ yarn add webpack webpack-cli --dev
+```
+
+Let's also create a file named `webpack.config.js` as well as a `src` folder and an empty `index.tsx` file in it.
+Lastly, I always set up an [.editorconfig](https://editorconfig.org/) file to let text editors sync on how to format things.
+
+A very basic configuration for Webpack doing nothing would be something like:
+
+```js
+const path = require("path");
+
+module.exports = {
+    context: __dirname,
+    mode: "development",
+    entry: {
+        app: "./src/index.tsx"
+    },
+    output: {
+        path: path.resolve(__dirname, "dist"),
+        filename: "app.[hash].js"
+    },
+    devtool: "eval-source-map",
+    resolve: {
+        extensions: [".js", ".jsx", ".ts", ".tsx"]
+    },
+    module: {
+        rules: []
+    },
+};
+```
+This means we are using `src/index.tsx` as the entry point (e.g. what Webpack is loading) and the output will be created in
+the `dist` directory, containing the hash in the filename for caching reasons.
+We use `eval-source-map` as `devtool` value as this is the best setting for development, You can see all the different values
+in the [Webpack devtool documentation](https://webpack.js.org/configuration/devtool/#devtool). Finally, we let Webpack know that
+it should try to load imports with no extensions as `.js`, `.jsx`, .`ts` and `.tsx` (so we can do `import something from "../index"`) and add a script to build our project.
+
+```diff
+diff --git a/package.json b/package.json
+index eb50375..1bc2df7 100644
+--- a/package.json
++++ b/package.json
+@@ -1,4 +1,8 @@
+ {
++  "scripts": {
++    "build": "webpack",
```

Comment:
In case you might be interested, here's the scripts I used

```json
  "scripts": {
    "build:dev": "webpack-dashboard -- webpack --watch",
    "build:stg": "webpack -p --env=staging --progress",
    "build:prod": "webpack -p --env=production --progress",
    "build:ci": "CI=1 yarn build:stg --bail --progress=0",
 }
```

Reply:
J'avais jamais vu webpack-dashboard avant, ca vaut le coup vs dev server avec stats=minimal?

Reply:
c'est super cool ouais. ca te permet de voir quels packages prennent le plus de poids dans tes chunks, et donc de mieux optimiser.

Reply:
Je vais parler de la taille du bundle dans la partie 2 ;)